symengine/series_generic.cpp

```diff
@@ -2,155 +2,278 @@
-        } );
-    poly_ = univariate_int_polynomial(var_, std::move(dict_trunc));
+int UnivariateSeries::compare(const Basic &other) const {
+    if (not is_a<UnivariateSeries>(other))
+        throw std::domain_error("cannot compare with UnivariateSeries");
+    const UnivariateSeries &o = static_cast<const UnivariateSeries &>(other);
+    return p_.get_basic()->__cmp__(*o.p_.get_basic());
 }
 
-UnivariateSeries::UnivariateSeries(const RCP<const Symbol> &var, const unsigned int &precision, const std::vector<integer_class> &v) :
-        var_{var}, prec_{precision} {
+bool UnivariateSeries::operator==(const UnivariateSeries &other) const {
+    return p_ == (other.p_);
+}
 
-    std::vector<integer_class> vtrunc;
-    std::copy_if(v.begin(), v.end(), std::back_inserter(vtrunc),
-        [&](decltype(v[0]) i) { return i < prec_; } );
-    poly_ = UnivariateIntPolynomial::create(var_, vtrunc);
+RCP<const Basic> UnivariateSeries::as_basic() const {
+    return p_.get_basic();
 }
 
-bool UnivariateSeries::is_canonical(const UnivariateIntPolynomial& poly, const unsigned int &prec) const
-{
-    return true;
+umap_int_basic UnivariateSeries::as_dict() const {
+    umap_int_basic map;
+    for (int i = 0; i <= degree_; i++) 
+       map[i] = p_.get_univariate_poly()->get_dict().at(i).get_basic();
+    return map;
 }
 
-std::size_t UnivariateSeries::__hash__() const
-{
-    return poly_->hash() + std::size_t(prec_ * 84728863L);
+RCP<const Basic> UnivariateSeries::get_coeff(int deg) const {
+    if(p_.get_univariate_poly()->get_dict().count(deg) == 0)
+        return zero;
+    else
+        return p_.get_univariate_poly()->get_dict().at(deg).get_basic();
 }
 
-bool UnivariateSeries::__eq__(const Basic &other) const
-{
-    if (not is_a<UnivariateSeries>(other))
-        return false;
-    const UnivariateSeries &o = static_cast<const UnivariateSeries &>(other);
-    return operator==(o);
+UnivariateExprPolynomial UnivariateSeries::var(const std::string &s) {
+    return UnivariateExprPolynomial(std::stoi(s));
 }
 
-int UnivariateSeries::compare(const Basic &other) const
-{
-    if (not is_a<UnivariateSeries>(other))
-        throw std::domain_error("cannot compare with UnivariateSeries");
-    const UnivariateSeries &o = static_cast<const UnivariateSeries &>(other);
-    return poly_->compare(*o.poly_);
+Expression UnivariateSeries::convert(const Number &x) {
+    return Expression(x.rcp_from_this());
+}
+
+RCP<const UnivariatePolynomial> UnivariateSeries::convert_poly(const map_int_Expr &d, unsigned pr) {
+    map_int_Expr dict_trunc;
+    unsigned int max = 0;
+    for (const auto &it : d) {
+        if ((unsigned)it.first < pr) {
+            if (max < (unsigned)it.first)
+                max = (unsigned)it.first;
+            dict_trunc[it.first] = it.second;
+        }
+    }
+    return univariate_polynomial(symbol(var_), max, std::move(dict_trunc));
 }
 
 std::string UnivariateSeries::__str__() const
 {
     std::ostringstream o;
```

Comment:
You can use the printer method from UnivariatePolynomial.
Also this belongs in printer.cpp